gdpr_assist/models.py

```diff
@@ -246,33 +278,278 @@ def _cast_class(cls, model, privacy_meta):
+                    res += "%sStarting recursive for %s #%s.\n" % (
+                    tabify(indent_level), l['model_name'], l['target_pk'])
+                    indent_level += 1
+
+                elif l["event"] == EventLog.EVENT_RECURSIVE_END:
+                    indent_level -= 1
+                    res += "%sEnding recursive for %s #%s.\n" % (tabify(indent_level), l['model_name'], l['target_pk'])
+
+                elif l["event"] == EventLog.EVENT_ALREADY_ANONYMISED:
+                    res += "%s%s #%s already anonymised.\n" % (tabify(indent_level), l['model_name'], l['target_pk'])
+
+                else:
+                    res += "%s%s #%s flat fields anonymised.\n" % (
+                    tabify(indent_level), l['model_name'], l['target_pk'])
+
+            return res
+        else:
+            return "%s is anonymised, but does not have matching logs." % self
+
+    @classmethod
+    def get_anonymization_tree(cls, prefix="", doprint=False, objs=[]):
+        """
+        Print the result of the nesting defined above for a given model.
+        Useful for sanity and loop checking. Example output:
+
+        Class:
+        |-> outcome_assessments = (OutcomeAssessment [set_field]):
+            |-> score
+            |-> comment
+        |-> class_polls = (ClassPoll [set_field]):
+            |-> poll = (Poll [fk]):
+                |-> title
+                |-> question
+                |-> pollchoice_set = (PollChoice [set_field]):
+                    |-> choice
+                |-> pollsession_set = (PollSession [set_field]):
+                    |-> outcome_assessments = (OutcomeAssessment [set_field]):
+                        |-> score
+                        |-> comment
+                    |-> pollresponse_set = (PollResponse [set_field]):
+                        |-> response
+
+        """
+        ABANDON_AFTER_N_LEVELS = 10
+        BASE_PREFIX = "    "
+        if len(prefix) > len(BASE_PREFIX) * ABANDON_AFTER_N_LEVELS:
+            return "ERROR: shouldn't go %s levels deep, check for loops." % ABANDON_AFTER_N_LEVELS
+        res = ''
+        if doprint:
+            res += cls.__name__ + ":\n"
+
+        privacy_meta_model = cls._privacy_meta
+
+        flat_fields = privacy_meta_model.fields
+        if flat_fields:
+            for f in flat_fields:
+                res += "%s|-> %s\n" % (prefix, f)
+
+        for fk in privacy_meta_model.fk_fields:
+
+            try:
+                f = getattr(cls, fk)
+                if hasattr(f, 'field'):
+                    f = f.field
+                elif hasattr(f, 'rel'):
+                    f = f.rel
+                child_model = f.related_model
+                res += "%s|-> %s = (%s [fk]):\n" % (prefix, fk, child_model.__name__)
+
+                res += child_model.get_anonymization_tree(prefix=BASE_PREFIX + prefix, doprint=False)
+            except AttributeError as e:
+                print(e)
```

Comment:
You probably need to define the `logger` object, but...

```suggestion
                logger.exception("A message that makes sense???  <-- not sure what would trigger this here?")
```

Reply:
Per above, yes, we need a logger. Pretty much, this would come up if a non-existent FK was put into that field. (typo, or putting a many to many here by accident)